### PR TITLE
Add lambda support to CHACallGraph

### LINKED ISSUE #1459: Support lambdas in CHA call graph construction
This is a notable gap in WALA's call graph construction algorithms for Java bytecode.  I haven't worked out the full algorithm but I don't think this should be too hard to support.  Need to somehow pattern match on the IR or bytecodes and model the class that gets generated for the lambda.  Not sure if we want to also update the `ClassHierarchy` in the process.

### PR DESCRIPTION
**Pull request by msridhar**
Fixes #1459 

To support lambdas, we re-use logic from the `LambdaMethodTargetSelector` to handle calls that use `invokedynamic`.  `CHACallGraph` only includes edges for methods reachable from given entrypoints, and our handling of lambdas adds new synthetic classes to the class hierarchy (representing the classes generated for lambdas).  So to ensure a complete call graph, we do a second pass after adding these synthetic classes to recompute call targets.  This is probably not the most efficient way to do things, but I'm not sure it's the bottleneck; if we need to optimize in the future we should do so based on profiling.

Note that the handling of lambdas here is very imprecise; every call site of a functional interface method will have an edge to every lambda of that functional interface type.  Still, it could be practically useful, e.g., when combined with other heuristic filtering of irrelevant code.

We add new tests, and also fix a couple other minor bugs (e.g., there was no edge before to the synthetic method that invokes class initializers, so it wasn't being processed properly).

### CODE REVIEW

File: cast/src/main/java/com/ibm/wala/cast/ipa/cha/CrossLanguageClassHierarchy.java [hunk 1]
```diff
@@ -276,5 +276,13 @@ public static CrossLanguageClassHierarchy make(AnalysisScope scope, ClassLoaderF
...
     return HashSetFactory.make();
   }
+
   /* END Custom change: unresolved classes */
+
+  @Override
+  public void clearCaches() {
+    for (IClassHierarchy cha : hierarchies.values()) {
+      cha.clearCaches();
+    }
```
Comment (liblit) [inline cast/src/main/java/com/ibm/wala/cast/ipa/cha/CrossLanguageClassHierarchy.java:286]: I'd prefer a tacit style here:

```suggestion
    hierarchies.values().forEach(IClassHierarchy::clearCaches);
```

OK to ignore this suggestion if you prefer the explicit loop.

File: core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java [hunk 2]
```diff
@@ -138,6 +145,28 @@ public void init(Iterable<Entrypoint> entrypoints) throws CancelException {
...
+    // to update call targets to include lambdas, we clear all call target caches, iterate through
+    // all call sites, and re-compute the targets.
+    // TODO optimize if needed
+    targetCache.clear();
+    cha.clearCaches();
+    for (CGNode n : this) {
+      for (CallSiteReference site : Iterator2Iterable.make(n.iterateCallSites())) {
+        Iterator<IMethod> methods = getOrUpdatePossibleTargets(n, site);
+        while (methods.hasNext()) {
+          IMethod target = methods.next();
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java:158]:
```suggestion
        for (IMethod target : Iterator2Iterable.make(getOrUpdatePossibleTargets(n, site))) {
```

File: core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java [hunk 3]
```diff
@@ -138,6 +145,28 @@ public void init(Iterable<Entrypoint> entrypoints) throws CancelException {
...
+          if (isRelevantMethod(target)) {
+            CGNode callee = getNode(target, Everywhere.EVERYWHERE);
+            if (callee == null) {
+              throw new RuntimeException(
+                  "should have already created CGNode for " + target.toString());
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java:163]:
```suggestion
                  "should have already created CGNode for " + target);
```

File: core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java [hunk 4]
```diff
@@ -146,6 +175,12 @@ public IClassHierarchy getClassHierarchy() {
     return cha;
   }
 
+  /**
+   * Cache of possible targets for call sites.
+   *
+   * <p>In the future, this cache could be keyed on (MethodReference,isDispatch) pairs to save space
+   * and possibly time, where isDispatch indicates whether the call site is a virtual dispatch.
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java:182]:
```suggestion
   * <p>In the future, this cache could be keyed on ({@link com.ibm.wala.types.MethodReference},
   * {@code isDispatch}) pairs to save space and possibly time, where {@code isDispatch} indicates
   * whether the call site is a virtual dispatch.
```

File: core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java [hunk 5]
```diff
@@ -279,7 +347,7 @@ private void closure() throws CancelException {
     while (!newNodes.isEmpty()) {
       CGNode n = newNodes.pop();
       for (CallSiteReference site : Iterator2Iterable.make(n.iterateCallSites())) {
-        Iterator<IMethod> methods = getOrUpdatePossibleTargets(site);
+        Iterator<IMethod> methods = getOrUpdatePossibleTargets(n, site);
         while (methods.hasNext()) {
           IMethod target = methods.next();
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/callgraph/cha/CHACallGraph.java:352]:
```suggestion
        for (IMethod target : Iterator2Iterable.make(getOrUpdatePossibleTargets(n, site))) {
```

File: core/src/main/java/com/ibm/wala/ipa/summaries/LambdaMethodTargetSelector.java [hunk 6]
```diff
@@ -99,6 +99,26 @@ public IMethod getCalleeTarget(CGNode caller, CallSiteReference site, IClass rec
     return base.getCalleeTarget(caller, site, receiver);
   }
 
+  /**
+   * Get the summary class for a lambda factory, if it has already been created.
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/summaries/LambdaMethodTargetSelector.java:103]:
```suggestion
   * Gets the summary class for a lambda factory, if it has already been created.
```

File: core/src/main/java/com/ibm/wala/ipa/cha/IClassHierarchy.java [hunk 7]
```diff
@@ -186,4 +186,10 @@ public interface IClassHierarchy extends Iterable<IClass> {
    * @throws IllegalArgumentException if c2 is null
    */
   boolean isAssignableFrom(IClass c1, IClass c2);
+
+  /**
+   * clear internal caches that may be invalidated by addition of new classes, e.g., a cache of the
+   * results of {@link #getPossibleTargets(MethodReference)}
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/cha/IClassHierarchy.java:192]:
```suggestion
   * Clear internal caches that may be invalidated by addition of new classes, e.g., a cache of the
   * results of {@link #getPossibleTargets(MethodReference)}.
```

File: core/src/main/java/com/ibm/wala/ipa/summaries/LambdaMethodTargetSelector.java [hunk 8]
```diff
@@ -99,6 +99,26 @@ public IMethod getCalleeTarget(CGNode caller, CallSiteReference site, IClass rec
...
+   *
+   * @param caller the caller node
+   * @param site the call site reference
+   * @return the summary class for the lambda factory, or null if it has not been created
```
Comment (liblit) [inline core/src/main/java/com/ibm/wala/ipa/summaries/LambdaMethodTargetSelector.java:107]:
```suggestion
   * @return the summary class for the lambda factory, or {@code null} if it has not been created
```

**Review by liblit — CHANGES_REQUESTED** [review]
Looks good. I suggest a few minor tweaks to improve code and Javadoc style.

**Comment by msridhar** [comment]
> Looks good. I suggest a few minor tweaks to improve code and Javadoc style.

Thanks for the detailed review!  All suggestions applied.

**Review by liblit — APPROVED** [review]

**Pull request merged**